### PR TITLE
Pr takeoff landing

### PR DESCRIPTION
**Pull request by tumbili**
@julianoes @Stifael @MaEtUgR This is the code for the automated takeoff and landing in all altitude controlled modes.

I made a fix in the landing detector so that it doesn't check the rc throttle stick when not in full manual mode.
This prevents the landing detector to detect in_air when the throttle stick is raised in altitude controlled modes.
If this works out well then we could remove the _landed_locked flag from this PR. I had to introduce it due to the problem with the landing detector.

### CODE REVIEW

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 1]
```diff
@@ -1456,6 +1518,15 @@ MulticopterPositionControl::task_main()
 			_alt_hold_engaged = false;
 		}
 
+		// assume really landed if landed, and if we have seen landed state for at least 0.5 seconds
+		bool landed_for_sure = _vehicle_land_detected.landed && _time_landed != 0 && (hrt_elapsed_time(&_time_landed) > 5e5);
+
+		// once we are sure that we are landed, then set the landed lock which prevents noise from the landed detector
+		// to interrupt the takeoff
+		if (landed_for_sure && !_landed_locked) {
```
Comment (tumbili) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @julianoes @MaEtUgR @Stifael Maybe we don't need this anymore.

File: src/modules/land_detector/MulticopterLandDetector.cpp [hunk 2]
```diff
@@ -172,7 +172,8 @@ bool MulticopterLandDetector::_get_landed_state()
 	// Check if user commands throttle and if so, report not landed based on
 	// the user intent to take off (even if the system might physically still have
 	// ground contact at this point).
-	if (_manual.timestamp > 0 && _manual.z > 0.15f && _ctrl_mode.flag_control_manual_enabled) {
```
Comment (tumbili) [inline src/modules/land_detector/MulticopterLandDetector.cpp]: This was causing the landing detector to detect in_air in position control when the drone was on the ground idling and the user was raising his throttle stick.

**Comment by LorenzMeier** [comment]
@julianoes Can you coordinate with @tumbili and @MaEtUgR how to get these all in and tested? I think its all super worthwhile stuff.

**Comment by tumbili** [comment]
@LorenzMeier Yeah, it's super nice if you can just switch to position control, put the throttle stick slightly above mid position and the drone does a takeoff to specified altitude. Once I had it working I spent batteries on just doing it over and over again :-) Just need to make sure now it's fool proof.

**Comment by mhkabir** [comment]
This is awesome! 

A question -  will this work in pure ALTCTL without any XY position ? I ask because for example optical flow is not available till around 30cm off the ground. 

Regards, 
Kabir

**Comment by tumbili** [comment]
@mhkabir yes, it also works for altctl

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 1 above)
Comment (julianoes) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: I agree. We need to trust the "landed" flag.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 3]
```diff
@@ -1517,23 +1588,45 @@ MulticopterPositionControl::task_main()
...
 			} else if (_control_mode.flag_control_manual_enabled
-				   && _vehicle_land_detected.landed) {
-				/* don't run controller when landed */
-				_reset_pos_sp = true;
-				_reset_alt_sp = true;
-				_do_reset_alt_pos_flag = true;
-				_mode_auto = false;
-				reset_int_z = true;
-				reset_int_xy = true;
+				   && (landed_for_sure || (_landed_locked && _acc_z_lp < 4.0f))) {
```
Comment (julianoes) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: Where does `4.0f` come from? Has this been used/tested before?

File: src/modules/mc_pos_control/mc_pos_control_params.c [hunk 4]
```diff
@@ -468,3 +468,36 @@ PARAM_DEFINE_FLOAT(MPC_ACC_HOR_MAX, 5.0f);
...
+PARAM_DEFINE_FLOAT(MPC_TK_T_THRESH, 0.6f);
+
+/**
+ * Relative takeoff altitude.
+ *
+ * The altitude relative to the pre-takeoff altitude the drone is going to target after a takeoff.
+ *
+ * @min 0.5
+ * @max 10.0
+ * @group Multicopter Attitude Control
```
Comment (julianoes) [inline src/modules/mc_pos_control/mc_pos_control_params.c]: There is already a param for the takeoff altitude for auto modes. Do we need an additional param for ALTCTL/POSCTL? If so, it should be named very clearly that it only affects ALTCTL/POSCTL.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 3 above)
Comment (tumbili) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @julianoes The original landing code introduced this. If you search for it in the same file you will find it. We could do a define.

File: src/modules/mc_pos_control/mc_pos_control_params.c (hunk 4 above)
Comment (tumbili) [inline src/modules/mc_pos_control/mc_pos_control_params.c]: @mhkabir @julianoes What do you think, do we need a new parameter?

**Comment by Stifael** [comment]
so I just tested it in sitl. 
There is a reset issue. For instance, if I am disarmed, but throttle up, this value is in fact saved and as soon as I arm, the vehicle shoots up. I guess it is the easiest just to set everything to zero if disarmer. I will look into that

**Comment by tumbili** [comment]
@Stifael Cool, thanks for testing!

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 5]
```diff
@@ -1301,6 +1333,40 @@ void MulticopterPositionControl::control_auto(float dt)
...
+	if (!_takeoff_jumped) {
+		// ramp thrust setpoint up
+		if (_vel(2) > -_params.tko_speed) {
+			_takeoff_thrust_sp += 0.5f * dt;
+			_vel_sp.zero();
+			_vel_prev.zero();
+
+		} else {
+			// copter has reached half of the takeoff speed
+			// we can assume we are off the ground, set takeoff target velocity
```
Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: why half the speed? it switches when vel(2) > tko_speed

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 6]
```diff
@@ -1301,6 +1333,40 @@ void MulticopterPositionControl::control_auto(float dt)
...
+			_thrust_int(2) = -math::constrain(_thrust_int(2), _params.thr_min, _params.thr_max);
+			_landed_locked = false;
+		}
+	}
+
+	if (_takeoff_jumped) {
+		// reset previous and current velocity setpoint, now the position controller can take over
+		_vel_sp(2) = -_params.tko_speed;
+		_vel_sp_prev(2) = -_params.tko_speed;
+	}
```
Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: why do we not set _vel_sp(2) = _vel(2), since that is the last velocity we had and is larger thant tko_speed (otherwise it would never saw a jumped_takeoff)? This prevents from dropping in altitude.
Same for _vel_sp_prev

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 1 above)
Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: what happens if the "if-statement" just checks if landed_for_sure is correct? does it need the !_landed_locked in the if statement?

**Comment by MaEtUgR** [comment]
@tumbili thanks for the pr, this worked great in my first sitl tests.
@Stifael I can confirm this bug, if you turn throttle up and down while disarmed and the arm it does a hop...

**Comment by tumbili** [comment]
@MaEtUgR Hm, it probably needs to check if it's disarmed. It thinks it should do a takeoff.

**Comment by tumbili** [comment]
@MaEtUgR @Stifael Come on, why would you ever raise the throttle stick when disarmed, please! :-)

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 3 above)
Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: this number is anyway rather strange: at hover, the _acc_z_lp is around 0, when accelerating upward, it is negative (z is positive pointing upward), and when going down the acceleration is limited anyway (I only can go down very slowly with small acceleration), I will never achieve an acceleration of 4.0f downwards. 
I also tested what the values are at landing, and they are sliglty negative (~ -0.17). Hence, this 4.0f has currently not really any effect.

**Comment by MaEtUgR** [comment]
@tumbili :) I think it's good that we found this. it's unexpected behavior.
let's say you have a spring loaded throttle... or you're just playing with the stick... or whatever

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 7]
```diff
@@ -1517,23 +1588,45 @@ MulticopterPositionControl::task_main()
...
+
+				if (!_takeoff_jumped) {
+					/* don't run controller when landed */
+					_reset_pos_sp = true;
+					_reset_alt_sp = true;
+					_mode_auto = false;
+					reset_int_xy = true;
+					_alt_hold_engaged = false;
+					_pos_hold_engaged = false;
+
```
Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: probably we could just add the reset_int_z here as well

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 5 above)
Comment (tumbili) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @Stifael This is a wrong comment from before, needs fixing

**Comment by tumbili** [comment]
@MaEtUgR It was just kidding...

**Comment by tumbili** [comment]
@Stifael @MaEtUgR Ping me once you finished, I'll do a second round.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 8]
```diff
@@ -1517,23 +1588,45 @@ MulticopterPositionControl::task_main()
...
+					_pos_hold_engaged = true;
+					_landed_locked = false;
+				}
 
 				_att_sp.roll_body = 0.0f;
 				_att_sp.pitch_body = 0.0f;
-				_att_sp.yaw_body = _yaw;
-				_att_sp.thrust = 0.0f;
-
+				R = matrix::Eulerf(_att_sp.roll_body, _att_sp.pitch_body, _att_sp.yaw_body);
```
Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @tumbili: I followed your suggestion to use matrix lib. I actually have a PR for Julians refactor mc_pos_control, and then everything should work with matrix. At some point we should do the same with attitude controller

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 9]
```diff
@@ -1517,23 +1588,45 @@ MulticopterPositionControl::task_main()
...
 				_att_sp.timestamp = hrt_absolute_time();
+				matrix::Quatf q_sp = R;
+				memcpy(&_att_sp.q_d[0], &q_sp._data, sizeof(_att_sp.q_d));
+				_att_sp.q_d_valid = true;
```
Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: being picky:
since R is not used, we can just write:
matrix::Quatf q_sp = matrix::Eulerf(...);

**Comment by Stifael** [comment]
I think that should do the fix with the resetting.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 3 above)
Comment (tumbili) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @Stifael The reason for this was to protect against free fall. Otherwise once the landed_locked flag is set in air falsely you have no chance of coming out again.
However, the real problem is that the accel value is not updated anymore once the the landed_locked flag is set. We need to fix that.

Comment (tumbili) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @julianoes I think @AndreasAntener came up with that number, he promised me that it's good.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 1 above)
Comment (tumbili) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @Stifael That just prevents it from setting it if  it's already set.

**Comment by LorenzMeier** [comment]
@tumbili Can you rebase this one? And a quick test flight would be much appreciated.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 1 above)
Comment (julianoes) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @tumbili do we really need yet another delay here? the land detection has that already.

Comment (tumbili) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @julianoes If the landing detector works well, we can remove that.

**Comment by julianoes** [comment]
Rebased in #6065.
